File: src/main/java/x/samantha/commandsjda/abs/CommandContext.java
package x.samantha.commandsjda.abs;

import net.dv8tion.jda.api.MessageBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import org.apache.lucene.index.DirectoryReader;
import org.apache.lucene.queryparser.classic.ParseException;
import org.apache.lucene.queryparser.classic.QueryParser;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.ScoreDoc;
import org.apache.lucene.search.TopDocs;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import x.samantha.commandsjda.managers.CommandManager;
import x.samantha.commandsjda.util.CommandSearchUtil;

import java.io.IOException;
import java.util.Arrays;

public class CommandContext {

    private static final Logger log = LoggerFactory.getLogger(CommandContext.class);

    public Guild guild;
    public TextChannel channel;
    public Member member;
    public User author;
    public Message message;
    public GuildMessageReceivedEvent rawEvent;

    public String trigger;
    public String[] args;
    public String rawArgs;
    public ICommand command;

    private CommandContext(Guild guild, TextChannel channel, Member member, User author, Message message, GuildMessageReceivedEvent rawEvent) {
        this.guild = guild;
        this.channel = channel;
        this.member = member;
        this.author = author;
        this.message = message;
        this.rawEvent = rawEvent;
    }

    /**
     * Create a CommandContext object
     * @return CommandContext
     */
    public static CommandContext parse(GuildMessageReceivedEvent e, String prefix, Boolean enableUnknownCommandMessage) {
        String input;
        String[] tempArgs;
        String msg = e.getMessage().getContentRaw();

        // Check if the message starts with the set prefix, return null if it doesn't
        if (msg.startsWith(prefix)) {
            input = msg.substring(prefix.length()).trim();
        } else {
            return null;
        }

        // If there is nothing after the prefix, return null, else continue
        if (input.isEmpty()) return null;
        tempArgs = input.split("\\s+");
        if (tempArgs.length == 0) return null;

        // Try to get the specified command from the command registry
        ICommand cmd = CommandManager.getCommand(tempArgs[0].toLowerCase());

        // If the command exists create a new CommandContext Object and return it
        if (cmd != null) {
            CommandContext ctx = new CommandContext(
                    e.getGuild(),
                    e.getChannel(),
                    e.getMember(),
                    e.getAuthor(),
                    e.getMessage(),
                    e
            );
            ctx.trigger = tempArgs[0];
            ctx.command = cmd;
            ctx.args = Arrays.copyOfRange(tempArgs, 1, tempArgs.length);
            ctx.rawArgs = input.replaceFirst(tempArgs[0], "");
            return ctx;
        } else {
            // If the command doesn't exist and enableUnknownCmdMsg is true then suggest similar commands
            if (enableUnknownCommandMessage) {
                try {
                    Query q = new QueryParser("command", CommandSearchUtil.analyzer).parse(tempArgs[0].toLowerCase());
                    int hitsPerPage = 5;
                    DirectoryReader reader = DirectoryReader.open(CommandSearchUtil.index);
                    IndexSearcher searcher = new IndexSearcher(reader);
                    TopDocs docs = searcher.search(q, hitsPerPage);
                    ScoreDoc[] hits = docs.scoreDocs;
                    MessageBuilder msgBuilder = new MessageBuilder("The Command `" + tempArgs[0].toLowerCase() + "` was not found, did you mean any of the following?\n```");
                    Arrays.stream(hits).forEach(hit -> {
                        try {
                            msgBuilder.append(searcher.doc(hit.doc).get("command").toLowerCase());
                        } catch (IOException ex2) {
                            log.warn("Unable to add command to suggestion list: " + ex2.getMessage());
                        }
                    });
                    reader.close();
                    msgBuilder.append("\n```");
                    e.getChannel().sendMessage(msgBuilder.build()).queue();
                } catch (ParseException | IOException ex) {
                    e.getChannel().sendMessage("Unknown Command").queue();
                    log.warn("Unknown Command '" + tempArgs[0].toLowerCase() + "' and unable to find similar commands: " + ex.getMessage());
                }
            }
            // Return null for commands that don't exist
            return null;
        }
    }

    Boolean hasArguments() {
        return args.length != 0 && !rawArgs.isEmpty();
    }
}

File: src/main/kotlin/x/samantha/commandsjda/abs/CommandContext.kt
package x.samantha.commandsjda.abs

import x.samantha.commandsjda.managers.CommandManager
import x.samantha.commandsjda.util.CommandSearchUtil
import net.dv8tion.jda.api.MessageBuilder
import net.dv8tion.jda.api.entities.*
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import org.apache.lucene.index.DirectoryReader
import org.apache.lucene.queryparser.classic.QueryParser
import org.apache.lucene.search.IndexSearcher

class CommandContext

private constructor(
    val guild: Guild,
    val channel: TextChannel,
    val member: Member,
    val author: User,
    val message: Message,
    val rawEvent: MessageReceivedEvent
) {

    lateinit var trigger: String
    lateinit var args: Array<String>
    lateinit var rawArgs: String
    lateinit var command: ICommand

    companion object {
        /**
         * Create a CommandContext object
         * @return CommandContext
         */
        fun parse(event: MessageReceivedEvent, prefix: String, enableUnknownCmdMsg: Boolean): CommandContext? {
            val input: String
            val tempArgs: Array<String>
            val message = event.message.contentRaw

            // Check if the message starts with the set prefix, return null if it doesn't
            if (message.startsWith(prefix)) {
                input = message.substring(prefix.length).trim { it <= ' ' }
            } else {
                return null
            }

            // If there is nothing after the prefix, return null, else continue
            if (input.isEmpty()) return null
            tempArgs = input.split("\\s+".toRegex()).toTypedArray()
            if (tempArgs.isEmpty()) return null

            // Try to get the specified command from the command registry
            val command = CommandManager.getCommand(tempArgs[0].toLowerCase())

            // If the command exists create a new CommandContext Object and return it
            if (command != null) {
                val ctx = CommandContext(
                        event.guild,
                        event.textChannel,
                        event.member!!,
                        event.author,
                        event.message,
                        event
                )
                ctx.trigger = tempArgs[0]
                ctx.command = command
                ctx.args = tempArgs.copyOfRange(1, tempArgs.size)
                ctx.rawArgs = input.replaceFirst(tempArgs[0], "")
                return ctx
            } else {
                // If the command doesn't exist and enableUnknownCmdMsg is true then suggest similar commands
                if (enableUnknownCmdMsg) {
                    val query = QueryParser("command", CommandSearchUtil.analyzer).parse(tempArgs[0].toLowerCase())
                    val hitsPerPage = 5
                    val reader = DirectoryReader.open(CommandSearchUtil.index)
                    val searcher = IndexSearcher(reader)
                    val docs = searcher.search(query, hitsPerPage)
                    val hits = docs.scoreDocs
                    val msgBuilder = MessageBuilder("The command `${tempArgs[0].toLowerCase()}` was not found, did you mean any of the following?\n```")
                    hits.forEach {
                        msgBuilder.append("${searcher.doc(it.doc).get("command").toLowerCase()}\n")
                    }
                    reader.close()
                    msgBuilder.append("\n```")
                    event.textChannel.sendMessage(msgBuilder.build()).queue()
                }
                // Return null for commands that don't exist
                return null
            }
        }
    }

    /**
     * Check if the message contained additional arguments other than the command trigger
     * @return boolean
     */
    fun hasArguments(): Boolean {
        return args.isNotEmpty() && !rawArgs.isEmpty()
    }
}
